jetty-http/src/main/java/org/eclipse/jetty/http/HttpParser.java

```diff
@@ -1052,18 +1050,29 @@ else if (_endOfContent == EndOfContent.CHUNKED_CONTENT)
                     case COOKIE:
                     case CACHE_CONTROL:
                     case USER_AGENT:
-                        addToConnectionTrie = _fieldCache != null && _field == null;
+                        addToFieldCache = _field == null;
                         break;
 
                     default:
                         break;
                 }
 
-                if (addToConnectionTrie && !_fieldCache.isFull() && _header != null && _valueString != null)
+                // Cache field?
+                if (addToFieldCache && _header != null && _valueString != null)
                 {
-                    if (_field == null)
-                        _field = new HttpField(_header, caseInsensitiveHeader(_headerString, _header.asString()), _valueString);
-                    _fieldCache.put(_field);
+                    if (_fieldCache == null)
+                    {
+                        _fieldCache = (_handler.getHeaderCacheSize() > 0 && (_version != null && _version == HttpVersion.HTTP_1_1))
+                            ? new ArrayTernaryTrie<>(_handler.getHeaderCacheSize())
```

Comment:
The `getHeaderCacheSize` method name is a bit misleading as the `ArrayTernaryTrie` is going to consume a lot more bytes in heap than the specified size, which doesn't match the `HttpConfiguration`'s managed attribute doc that reads _"The maximum allowed size in bytes for an HTTP header field cache"_.

Reply:
updated javadoc